### PR TITLE
[asan] Switch initialization to "double-checked locking"

### PR DESCRIPTION
**Pull request by vitalybuka**
This allows to remove `asan_init_is_running` which likely had a data
race.

Simplifies https://github.com/llvm/llvm-project/pull/74086 and reduces a difference between platforms.

### CODE REVIEW

File: compiler-rt/lib/asan/asan_rtl.cpp [hunk 1]
```diff
@@ -408,9 +406,8 @@ static void AsanInitInternal() {
   // Stop performing init at this point if we are being loaded via
   // dlopen() and the platform supports it.
   if (SANITIZER_SUPPORTS_INIT_FOR_DLOPEN && UNLIKELY(HandleDlopenInit())) {
-    SetAsanInitIsRunning(0);
     VReport(1, "AddressSanitizer init is being performed for dlopen().\n");
-    return;
+    return false;
```
Comment (dvyukov) [inline compiler-rt/lib/asan/asan_rtl.cpp:410]: If we return false here, __asan_handle_no_return won't unpoison stack.
It looks like it's different from the current behavior.
What's the reason for the change?

File: compiler-rt/lib/asan/asan_rtl.cpp [hunk 2]
```diff
@@ -515,22 +511,27 @@ static void AsanInitInternal() {
...
+    return;
+  SpinMutexLock lock(&asan_inited_mutex);
+  AsanInitInternal();
 }
 
 bool TryAsanInitFromRtl() {
-  if (UNLIKELY(AsanInitIsRunning()))
+  if (LIKELY(AsanInited()))
+    return true;
+  if (!asan_inited_mutex.TryLock())
```
Comment (dvyukov) [inline compiler-rt/lib/asan/asan_rtl.cpp:530]: This also seems to introduce a change in the behavior. Currently __asan_handle_no_return returns early only while AsanInitIsRunning, but the mutex lock duration covers whole AsanInitInternal.
Is this intentional?

File: compiler-rt/lib/asan/asan_rtl.cpp (hunk 1 above)
Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:410]: We didn't get to InitializeShadowMemory below, seems now it must do nothing or crash anyway.

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:410]: Maybe it's OK to block and wait in __asan_handle_no_return, but as-is seems close to the current behavior.

File: compiler-rt/lib/asan/asan_rtl.cpp (hunk 2 above)
Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:530]: Actually I don't believe we care about partially initialized asan for `__asan_handle_no_return`.
I think `__asan_handle_no_return` should bailout aggressively.

File: compiler-rt/lib/asan/asan_rtl.cpp (hunk 1 above)
Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:410]: I believe `__asan_handle_no_return` is unsupported before initialization, so 
```
if (UNLIKELY(!AsanInited()))
    return;
```
should be as good as the current state.

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:410]: To clarify

Before the patch
* `asan_inited = 0, asan_init_is_running = 0`: `__asan_handle_no_return` proceeds but likely going to crash, e.g. on [`llvm-project/compiler-rt/lib/asan/asan_poisoning.cpp:41`](https://github.com/llvm/llvm-project/blob/43455a2f0d3cdcd8dc46c4fbdebf4e655b26c2f2/compiler-rt/lib/asan/asan_poisoning.cpp#L41)
So we don't see this case in practice.

* `asan_inited = 0, asan_init_is_running = 1`: `__asan_handle_no_return` returns doing nothing 

* `asan_inited = 1, asan_init_is_running = 0`: `__asan_handle_no_return` works as expected

After the patch:
* `!AsanInited()`: `__asan_handle_no_return` return doing noting

* `AsanInited()`: `__asan_handle_no_return` works as expected

File: compiler-rt/lib/asan/asan_rtl.cpp [hunk 3]
```diff
@@ -515,22 +511,27 @@ static void AsanInitInternal() {
...
 // Initialize as requested from some part of ASan runtime library (interceptors,
 // allocator, etc).
 void AsanInitFromRtl() {
-  CHECK(!AsanInitIsRunning());
-  if (UNLIKELY(!AsanInited()))
-    AsanInitInternal();
+  if (LIKELY(AsanInited()))
+    return;
+  SpinMutexLock lock(&asan_inited_mutex);
+  AsanInitInternal();
```
Comment (zacklj89) [inline compiler-rt/lib/asan/asan_rtl.cpp:524]: Does it make sense to wrap in `CHECK`?

**Review by zacklj89 — APPROVED** [review]

File: compiler-rt/lib/asan/asan_rtl.cpp (hunk 3 above)
Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:524]: We will have to do something about SANITIZER_SUPPORTS_INIT_FOR_DLOPEN
I'd like to avoid dealing with that in this patch.

**Review by dvyukov — APPROVED** [review]
I don't have any objections.
There may be some subtle implications, or maybe not. I guess it's impossible to understand w/o merging this.
The increased critical section may lead to deadlocks on recursion, or not.
